keps/sig-api-machinery/3285-subresource-finalizers/README.md

```diff
@@ -0,0 +1,784 @@
+`Role` and `ClusterRole` can be set to the endpoint, therefore users can set a separate RBAC to `/{resource_name}/finalizers`.
+A PoC implementation can be found [here](https://github.com/mkimuram/kubernetes/commit/751a67e03d62397c1d21947e9e2e3a81f279a29f) (Please also see [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) for how it works. Note that API definition in this KEP will differ from the one in the PoC).
+
+### API Changes
+
+A new `Finalize` resource is introduced in `subresources` API group:
+
+```golang
+// Finalize represents a finalizers field to be accessed via subresource.
+type Finalize struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Only the `finalizers` field is updateable.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+```
+
+### Supported API calls to the endpoint
+
+The `/{resource_name}/finalizers` endpoint supports 3 verbs, GET, PATCH, and PUT.
+
+#### GET
+
+GET verb is used to read the value in `metadata.finalizers` field.
+
+- Request format:
+Empty string is passed as a request.
+
+- Response format:
+  - On success, the `Finalize` resource is returned.
+    ex)
+    ```json
+    {
+        "kind": "Finalize",
+        "apiVersion": "v1alpha1",
+        "metadata": {
+            "finalizers": ["example.com/my-finalizer"],
+            "creationTimestamp": null,
+		}
+    }
+    ```
+
+  - On failure, error message is stored in `Status` resource.
+    ex)
+    ```json
+    {
+        "kind": "Status",
+        "apiVersion": "v1",
+        "metadata": {},
+        "status": "Failure",
+        "message": "\"finalize\" is forbidden: User \"example-user\" cannot get subresource \"finalizers\" of \"example\" in API group \"example\" in the namespace \"example\"",
+        "reason": "Forbidden",
+        "details": {},
+        "code": 403
+    }
+    ```
+
+- Status code:
+  - 200 "OK": Returned on success without change
+  - 403 "Forbidden": Returned when GET verb is not allowed for `/{resource_name}/finalizers` endpoint (RBAC admission handles this)
+
+#### PATCH
+
+PATCH verb is used to modify the value in `metadata.finalizers` field.
+
+- Request format:
+
+ex)
+```json
+{
+    "metadata":{
+        "finalizers": ["example.com/my-finalizer"]
+    }
+}
+```
+
+- Response format:
+  - On success, the `Finalize` resource is returned (the same format as the one for GET).
+  - On failure, error message is stored in `Status` resource (the same format as the one for GET)
+
+- Status code:
+  - 200 "OK": Returned on success without change
+  - 201 "Created": Returned on success with change
+  - 401 "Unauthorized": Returned on failure in modifying the value
+  - 403 "Forbidden": Returned when PATCH verb is not allowed for `/{resource_name}/finalizers` endpoint (RBAC admission handles this)
+
+#### PUT
+
+PUT verb is used to set the value in `metadata.finalizers` field.
```

Comment:
For test-cases: make sure PATCH and GET properly support precoditions on UID and RV

Reply:
Added in e2e test.